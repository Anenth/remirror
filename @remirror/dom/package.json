{
  "name": "@remirror/dom",
  "version": "0.0.0",
  "description": "Use remirror directly in the dom.",
  "keywords": [
    "remirror",
    "extension"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/next/@remirror/dom",
  "repository": "https://github.com/remirror/remirror/tree/master/@remirror/dom",
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "sideEffects": false,
  "main": "dist/dom.cjs.js",
  "module": "dist/dom.esm.js",
  "types": "dist/dom.cjs.d.ts",
  "files": [
    "dist"
  ],
  "dependencies": {
    "@babel/runtime": "^7.10.4",
    "@remirror/preset-core": "^0.0.0",
    "type-fest": "^0.16.0"
  },
  "devDependencies": {
    "@remirror/core": "^0.11.0",
    "@remirror/pm": "^0.0.0"
  },
  "peerDependencies": {
    "@remirror/core": "^0.11.0",
    "@remirror/pm": "^0.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "meta": {
    "sizeLimit": "5 KB"
  }
}
