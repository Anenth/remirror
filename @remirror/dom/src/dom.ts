import {
  AnyCombinedUnion,
  BuiltinPreset,
  EditorState,
  EditorWrapper,
  EditorWrapperOutput,
  EditorWrapperProps,
  EMPTY_PARAGRAPH_NODE,
  FromToParameter,
  getDocument,
  isFunction,
  RemirrorContentType,
  RemirrorManager,
  SchemaFromCombined,
  UpdateStateParameter,
} from '@remirror/core';
import { EditorView } from '@remirror/pm/view';
import { CorePreset, createCoreManager, CreateCoreManagerOptions } from '@remirror/preset-core';

/**
 * Create an editor manager. It comes with the `CorePreset` already available.
 */
export function createDomManager<Combined extends AnyCombinedUnion>(
  combined: Combined[],
  options?: CreateCoreManagerOptions,
): RemirrorManager<CorePreset | BuiltinPreset | Combined> {
  return createCoreManager([...combined], options);
}

export interface DomEditorWrapperProps<Combined extends AnyCombinedUnion>
  extends EditorWrapperProps<Combined> {
  /**
   * Provide a container element. Which the editor will be appended to.
   */
  element: Element;
}

/**
 * Create the prosemirror editor for the dom environment.
 */
export function createDomEditor<Combined extends AnyCombinedUnion>(
  props: DomEditorWrapperProps<Combined>,
): DomEditorWrapperOutput<Combined> {
  const { stringHandler, onError, manager, forceEnvironment, element } = props;

  function createStateFromContent(
    content: RemirrorContentType,
    selection?: FromToParameter,
  ): EditorState<SchemaFromCombined<Combined>> {
    return manager.createState({
      content,
      doc: getDocument(forceEnvironment),
      stringHandler,
      selection,
      onError,
    });
  }

  const fallback = isFunction(onError) ? onError() : onError ?? EMPTY_PARAGRAPH_NODE;
  const initialEditorState = createStateFromContent(props.initialContent ?? fallback);
  const wrapper = new DomEditorWrapper<Combined>({
    createStateFromContent,
    getProps: () => props,
    initialEditorState,
    element,
  });

  return wrapper.output;
}

interface DomEditorWrapperOutput<Combined extends AnyCombinedUnion>
  extends EditorWrapperOutput<Combined> {
  /**
   * Call this to cleanup the view.
   */
  destroy: () => void;
}

/**
 * The helper class which makes integrating with the DOM easier.
 */
class DomEditorWrapper<Combined extends AnyCombinedUnion> extends EditorWrapper<
  Combined,
  DomEditorWrapperProps<Combined>
> {
  /**
   * Create the prosemirror EditorView.
   */
  protected createView(state: EditorState<SchemaFromCombined<Combined>>, element?: Element) {
    return new EditorView(element, {
      state,
      nodeViews: this.manager.store.nodeViews,
      dispatchTransaction: this.dispatchTransaction,
      attributes: () => this.getAttributes(),
      editable: () => {
        return this.props.editable ?? true;
      },
    });

    this.onChange();
    this.addFocusListeners();
  }

  /**
   * Responsible for managing state updates.
   */
  protected updateState(parameter: UpdateStateParameter<SchemaFromCombined<Combined>>): void {
    const { state, tr, triggerChange = true } = parameter;

    // Update the internal prosemirror state. This happens before we update
    // the component's copy of the state.
    this.view.updateState(state);

    if (triggerChange) {
      this.onChange({ state, tr });
    }

    this.manager.onStateUpdate({ previousState: this.previousState, state });
  }

  get output() {
    return {
      ...this.editorWrapperOutput,
      destroy: () => this.onDestroy(),
    };
  }
}
