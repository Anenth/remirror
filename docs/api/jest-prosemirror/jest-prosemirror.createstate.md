<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [jest-prosemirror](./jest-prosemirror.md) &gt; [createState](./jest-prosemirror.createstate.md)

## createState variable

Create the editor state for a tagged prosemirror doc

<b>Signature:</b>

```typescript
createState: <GSchema extends EditorSchema<string, string> = any>(taggedDoc: pm.TaggedProsemirrorNode<GSchema>, plugins?: Plugin<any, any>[]) => EditorState<GSchema>
```
