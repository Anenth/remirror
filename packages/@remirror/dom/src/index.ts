export type { DomEditorWrapperProps } from './dom';
export { createDomEditor, createDomManager } from './dom';
