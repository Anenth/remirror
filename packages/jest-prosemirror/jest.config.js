const config = require('../../support/jest/jest.config');

module.exports = {
  ...config,
  setupFilesAfterEnv: [],
  snapshotSerializers: [],

  displayName: { name: require('./package.json').name.replace('@remirror/', ''), color: 'pink' },
};
