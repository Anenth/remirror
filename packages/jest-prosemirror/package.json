{
  "name": "jest-prosemirror",
  "version": "0.8.3",
  "description": "Write expressive tests for your prosemirror editor",
  "keywords": [
    "jest",
    "prosemirror",
    "snapshots",
    "tests"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/next/packages/jest-prosemirror",
  "repository": "https://github.com/remirror/remirror/tree/master/packages/jest-prosemirror",
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "main": "dist/jest-prosemirror.cjs.js",
  "module": "dist/jest-prosemirror.esm.js",
  "types": "dist/jest-prosemirror.cjs.d.ts",
  "files": [
    "dist",
    "environment",
    "serializer"
  ],
  "dependencies": {
    "@babel/runtime": "^7.10.1",
    "@remirror/pm": "^0.0.0",
    "@testing-library/dom": "^7.6.0",
    "@types/prosemirror-schema-basic": "^1.0.1",
    "@types/prosemirror-test-builder": "^1.0.1",
    "chalk": "^4.0.0",
    "jest-matcher-utils": "^26.0.1",
    "prosemirror-schema-basic": "^1.1.2",
    "prosemirror-tables": "^1.1.0",
    "prosemirror-test-builder": "^1.0.3",
    "remirror": "^0.13.1",
    "test-keyboard": "^0.7.6"
  },
  "devDependencies": {
    "jest": "^26.0.1"
  },
  "peerDependencies": {
    "jest": ">= 24"
  },
  "publishConfig": {
    "access": "public"
  },
  "preconstruct": {
    "entrypoints": [
      ".",
      "environment",
      "serializer"
    ]
  }
}
