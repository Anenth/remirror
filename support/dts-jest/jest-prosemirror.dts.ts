import { EventType } from 'jest-prosemirror';
import { EventType as DomEventType } from '@testing-library/dom';

const eventType: EventType = 'abort';
const domEventType: DomEventType = 'animationEnd';

const excludeEvents = (): Exclude<DomEventType, EventType> => undefined as never;
const excludeDomEvents = (): Exclude<EventType, DomEventType> => undefined as never;

// @dts-jest:pass:snap
excludeEvents();

// @dts-jest:pass:snap
excludeDomEvents();
