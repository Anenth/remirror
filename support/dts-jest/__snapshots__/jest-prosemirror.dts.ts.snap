// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`excludeDomEvents() (type) should match snapshot 1`] = `"\\"tripleClick\\""`;

exports[`excludeEvents() (type) should match snapshot 1`] = `"\\"popState\\""`;
